file.go

```diff
@@ -0,0 +1,217 @@
+	// so an initial read of 1 byte would not work correctly.
+	if size < 512 {
+		size = 512
+	}
+
+	data := make([]byte, 0, size)
+	for {
+		if time.Now().After(t) {
+			return data, mtime, os.ErrDeadlineExceeded
+		}
+		if len(data) >= cap(data) {
+			d := append(data[:cap(data)], 0)
+			data = d[:len(data)]
+		}
+		n, err := f.Read(data[len(data):cap(data)])
+		data = data[:len(data)+n]
+		if err != nil {
+			if err == io.EOF {
+				err = nil
+			}
+			return data, mtime, err
+		}
+	}
+}
+
+var (
+	mtimeName = "expexp_file_mtime"
+	mtimeHelp = "Time of modification of parsed file (in miliseconds)"
+	mtimeType = dto.MetricType_GAUGE
+	mtimeLabelModule = "module"
+	mtimeLabelPath   = "path"
+)
+
+func (c fileConfig) GatherWithContext(ctx context.Context, r *http.Request, path string) prometheus.GathererFunc {
+	return func() ([]*dto.MetricFamily, error) {
+
+		errc := make(chan error, 1)
+		datc := make(chan []byte, 1)
+		timec := make(chan time.Time, 1)
+		go func() {
+			deadline, ok := ctx.Deadline()
+			if ! ok { deadline = time.Now().Add(time.Minute * 5) }
+			dat, mtime, err := readFileWithDeadline(path, deadline)
+			errc <- err
+			if err == nil {
+			    datc <- dat
+			    timec <- mtime
+			}
+			close(errc)
+			close(datc)
+			close(timec)
+		}()
+
+		err := <- errc
+		if err != nil {
+			log.Warnf("File module %v failed to read file %v, %+v", c.mcfg.name, path, err)
+			fileFailsCount.WithLabelValues(c.mcfg.name).Inc()
+			if err == context.DeadlineExceeded || err == os.ErrDeadlineExceeded {
+				proxyTimeoutCount.WithLabelValues(c.mcfg.name).Inc()
+			}
+			return nil, err
+		}
+		dat := <- datc
+		mtime := <- timec
+		var prsr expfmt.TextParser
+
+		var mtimeBuf *int64 = nil
+		if ! mtime.IsZero() {
+			mtimeBuf = new(int64)
+			*mtimeBuf = mtime.UnixMilli()
+		}
+
+		var result []*dto.MetricFamily
+		mfs, err := prsr.TextToMetricFamilies(bytes.NewReader(dat))
+		if err != nil {
+			proxyMalformedCount.WithLabelValues(c.mcfg.name).Inc()
+			return nil, err
+		}
+		for _, mf := range mfs {
+			if c.UseMtime && mtimeBuf != nil {
+				for _, m := range mf.GetMetric() {
+					m.TimestampMs = mtimeBuf
+				}
+			}
+			result = append(result, mf)
+		}
+		if !mtime.IsZero() {
+			v := float64(mtime.UnixMilli())
+			g := dto.Gauge { Value: &v, }
+			l := make([]*dto.LabelPair, 2)
+			l[0] = &dto.LabelPair{
+				Name:&mtimeLabelModule,
+				Value:&c.mcfg.name,
+			}
+			l[1] = &dto.LabelPair{
+				Name:&mtimeLabelPath,
+				Value:&path,
+			}
+			m := dto.Metric {
+				Label: l,
+				Gauge: &g,
+			}
+			mf := dto.MetricFamily{
+				Name: &mtimeName,
+				Help: &mtimeHelp,
+				Type: &mtimeType,
+			}
+			mf.Metric = append(mf.Metric, &m)
+			result = append(result, &mf)
+		}
+		return result, nil
+	}
+}
+
+var cleanSlashes = regexp.MustCompile("(^|/)/+")
+
+func (c fileConfig) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	qvs := r.URL.Query()
+	path := cleanSlashes.ReplaceAllString(qvs.Get("path"),"$1")
```

Comment:
I think this change is acceptable if the file paths are set in the config, but am not willing to accept them as part of the URL. This opens a serious can of worms in terms of potential security implications.

Reply:
Also, can't you already achieve what you want (minus the external file selection), with an `exec` module and a shell script?

Reply:
Shell scripts are in general more dangerous than regexps, because they rely on environment variables, state of filesystem and many more things, so that the surface for attack is much larger. My motivation was to enable similar functionality to passing query parameters in http module. (Passing query parameters to http exporters can be also dangerous, moreover you do not have any validation mechanism, to let operator which query parameters and values are allowed.)

That being said, I will remove path parameter, because solution, that we are designing at my work, can handle individual module configuration file for each allowed input file (although it will create some possibilities for misconfigurations). I will change timestamp to seconds as requested and rebase the branch, so that the code handling the query parameter will not be part of (history of) merge request.